tgui/packages/tgui/interfaces/Biogenerator.js

```diff
@@ -0,0 +1,247 @@
+  } = data;
+
+  return (
+    <Section
+      title="Storage">
+      <Flex>
+        <Flex.Item mr="20px" color="silver">
+          Biomass:
+        </Flex.Item>
+        <Flex.Item mr="5px">
+          {biomass}
+        </Flex.Item>
+        <Icon
+          name="leaf"
+          size={1.2}
+          color="#3d8c40"
+        />
+      </Flex>
+      <Flex mt="8px">
+        <Flex.Item mr="10px" color="silver">
+          Container:
+        </Flex.Item>
+          <ProgressBar
+            value={container_curr_reagents}
+            maxValue={container_max_reagents}>
+            <Box textAlign="center">
+              {container_curr_reagents + ' / ' + container_max_reagents + ' units'}
+            </Box>
+          </ProgressBar>
+      </Flex>
+    </Section>
+  );
+};
+
+const Controls = (props, context) => {
+  const { act, data } = useBackend(context);
+  const { has_plants } = data;
+
+  return (
+    <Section title="Controls">
+      <Flex>
+        <Flex.Item width="30%" mr="3px">
+          <Button
+            fluid
+            textAlign="center"
+            icon="power-off"
+            disabled={!has_plants}
+            tooltip={has_plants ? "" : "There are no plants in the biogenerator."}
+            tooltipPosition="top-right"
+            content="Activate"
+            onClick={() => act('activate')}
+          />
+        </Flex.Item>
+        <Flex.Item width="40%" mr="3px">
+          <Button
+            fluid
+            textAlign="center"
+            icon="flask"
+            content="Detach Container"
+            onClick={() => act('detach_container')}
+          />
+        </Flex.Item>
+        <Flex.Item width="30%">
+          <Button
+            fluid
+            textAlign="center"
+            icon="eject"
+            disabled={!has_plants}
+            tooltip={has_plants ? "" : "There are no stored plants to eject."}
+            tooltipPosition="top-left"
+            content="Eject Plants"
+            onClick={() => act('eject_plants')}
+          />
+        </Flex.Item>
+      </Flex>
+    </Section>
+  );
+};
+
+const Products = (props, context) => {
+  const { act, data } = useBackend(context);
+  const {
+    biomass,
+    product_list,
+    efficiency,
+  } = data;
+
+  let [vendAmount, setVendAmount] = useSharedState(context, 'vendAmount', 1);
+
+  let content = Object.entries(product_list).map((kv, _i) => {
+    let category_items = Object.entries(kv[1]).map((kv2) => {
+      return kv2[1];
+    })
+
+    return (
+      <Collapsible
+        key={kv[0]}
+        title={kv[0]}
+        open>
+        {category_items.map((item) => (
+          <Flex
+            key={item}
+            py="2px"
+            className="candystripe"
+            align="center">
+            <Flex.Item width="50%" ml={0.5}>
+              {item.name}
+            </Flex.Item>
+            <Flex.Item textAlign="right" width="20%">
+              {(item.cost / efficiency) * vendAmount}
```

Comment:
i think i would prefer if js didn't care about efficiency and doing the math, and dm passed in an already efficiency-adjusted cost. Less places to do math means less places to adjust when math does need changing.
Oh, and if you store the already adjusted prices in the product_list, then it would make even dm side math easier - `check_cost` would become trivial and inlineable as an example.

Reply:
Yeah I don't know why I decided to do that on the FE. Makes much more sense to do the calculators and store it in DM.